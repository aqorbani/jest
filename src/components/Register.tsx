import { ChangeEvent, useState } from "react";
import { useNavigate } from "react-router-dom";
import { EMAIL_PATTERN, ERROR_MESSAGE } from "../constants/validation";

export default function Register() {
  const [registerInput, setRegisterInput] = useState({
    email: "",
    password: "",
    confirmPassword: "",
  });

  const [error, setError] = useState("");

  const navigate = useNavigate();

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setRegisterInput({ ...registerInput, [e.target.name]: e.target.value });
  };

  const submitHandler = () => {
    if (!EMAIL_PATTERN.test(registerInput.email)) {
      return setError(ERROR_MESSAGE.EMAIL);
    } else if (registerInput.password.length < 5) {
      return setError(ERROR_MESSAGE.PASSWORD);
    } else if (registerInput.password !== registerInput.confirmPassword) {
      return setError(ERROR_MESSAGE.CONFIRM_PASSWORD);
    }
    navigate("/products");
  };

  return (
    <div>
      <h1>Register</h1>
      {error && <p className="text-red-800 font-semibold">{error}</p>}
      <div>
        <label htmlFor="email">ایمیل</label>
        <input
          type="email"
          id="email"
          name="email"
          value={registerInput.email}
          onChange={handleChange}
        />
      </div>
      <div>
        <label htmlFor="password">رمز عبور</label>
        <input
          type="password"
          id="password"
          name="password"
          value={registerInput.password}
          onChange={handleChange}
        />
      </div>
      <div>
        <label htmlFor="confirmPassword">تکرار رمز عبور</label>
        <input
          type="confirmPassword"
          id="confirmPassword"
          name="confirmPassword"
          value={registerInput.confirmPassword}
          onChange={handleChange}
        />
      </div>
      <div>
        <button type="submit" onClick={submitHandler}>
          ثبت نام
        </button>
      </div>
    </div>
  );
}
