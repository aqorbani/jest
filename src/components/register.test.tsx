import { render, screen } from "@testing-library/react";
import Register from "./Register";
import { BrowserRouter } from "react-router-dom";

const getElement = (element: string) => {
  interface Obj {
    "email": HTMLInputElement;
    "password": HTMLInputElement;
    "confPassword": HTMLInputElement;
  }
  const elements: Obj = {
    "email": screen.getByRole("textbox", { name: "ایمیل" }),
    "password": screen.getByLabelText("رمز عبور"),
    "confPassword": screen.getByLabelText("تکرار رمز عبور"),
  };
  if (elements[element as keyof Obj]) return elements[element as keyof Obj];
};

describe("Register page", () => {
  test("Input should be initially empty", () => {
    // ARRANGE
    render(
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    );

    // ACT
    // change event - click event or ...
    // ASSERT

    expect(getElement("email").value).toBe("");
    expect(getElement("password").value).toBe("");
    expect(getElement("confPassword").value).toBe("");

    // getElement
    // findElement
    // queryElement
    // getAll...
    // finaAll...
    // queryAll...
  });
});
