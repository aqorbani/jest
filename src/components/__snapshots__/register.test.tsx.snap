// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register page Handle errors and navigation snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="p-2 flex flex-col justify-center items-center"
      >
        <h3
          class="font-extrabold text-2xl"
        >
          ثبت نام
        </h3>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="email"
          >
            ایمیل
          </label>
          <input
            class="w-full"
            id="email"
            name="email"
            placeholder="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="password"
          >
            رمز عبور
          </label>
          <input
            class="w-full"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="confirmPassword"
          >
            تکرار رمز عبور
          </label>
          <input
            class="w-full"
            id="confirmPassword"
            name="confirmPassword"
            type="confirmPassword"
            value=""
          />
        </div>
        <div
          class="w-80"
        >
          <button
            class="w-full bg-orange-900 rounded text-white"
            type="submit"
          >
            ثبت نام
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="p-2 flex flex-col justify-center items-center"
      >
        <h3
          class="font-extrabold text-2xl"
        >
          ثبت نام
        </h3>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="email"
          >
            ایمیل
          </label>
          <input
            class="w-full"
            id="email"
            name="email"
            placeholder="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="password"
          >
            رمز عبور
          </label>
          <input
            class="w-full"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="w-80 m-1 bg-orange-300 p-2 rounded"
        >
          <label
            for="confirmPassword"
          >
            تکرار رمز عبور
          </label>
          <input
            class="w-full"
            id="confirmPassword"
            name="confirmPassword"
            type="confirmPassword"
            value=""
          />
        </div>
        <div
          class="w-80"
        >
          <button
            class="w-full bg-orange-900 rounded text-white"
            type="submit"
          >
            ثبت نام
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="p-2 flex flex-col justify-center items-center"
    >
      <h3
        class="font-extrabold text-2xl"
      >
        ثبت نام
      </h3>
      <div
        class="w-80 m-1 bg-orange-300 p-2 rounded"
      >
        <label
          for="email"
        >
          ایمیل
        </label>
        <input
          class="w-full"
          id="email"
          name="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="w-80 m-1 bg-orange-300 p-2 rounded"
      >
        <label
          for="password"
        >
          رمز عبور
        </label>
        <input
          class="w-full"
          id="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="w-80 m-1 bg-orange-300 p-2 rounded"
      >
        <label
          for="confirmPassword"
        >
          تکرار رمز عبور
        </label>
        <input
          class="w-full"
          id="confirmPassword"
          name="confirmPassword"
          type="confirmPassword"
          value=""
        />
      </div>
      <div
        class="w-80"
      >
        <button
          class="w-full bg-orange-900 rounded text-white"
          type="submit"
        >
          ثبت نام
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
